deeplake/core/chunk_engine.py

```diff
@@ -2495,12 +2495,139 @@ def _sequence_item_length_range(self):
+                                    )
+                                    or (1,)
+                                )
+                else:
+                    self.check_link_ready()
+                    shape = self.read_shape_for_sample(idx)  # type: ignore
+                    # if link verification was not done
+                    if len(shape) > sample_ndim:
+                        sample_ndim = len(shape)
+                        sample_shapes = np.zeros(
+                            (num_samples, sample_ndim), dtype=np.int32
+                        )
+                if flatten:
+                    start, end = self.sequence_encoder[i]
+                    sample_shapes[start:end] = shape
+                else:
+                    sample_shapes[i] = shape
+        return sample_shapes
+
+    def shapes(
+        self,
+        index: Index,
+        sample_shape_provider: Optional[Callable] = None,
+        pad_tensor: bool = False,
+    ):
+        tensor_ndim = self.ndim()
+
+        if len(index) > 1:
+            raise IndexError(f"`.shapes` only accepts indexing on the primary axis.")
+
+        index_0 = index.values[0]
+        if self.is_sequence:
+            sample_indices = list(index_0.indices(self._sequence_length))
+            num_samples = sum(
+                map(
+                    lambda x: x[1] - x[0],
+                    [self.sequence_encoder[i] for i in sample_indices],
+                )
+            )
+            sample_ndim = tensor_ndim - 2
+        else:
+            num_samples = index_0.length(self.num_samples)
+            sample_ndim = tensor_ndim - 1
+
+        sample_shapes = np.zeros((num_samples, sample_ndim), dtype=np.int32)
+
+        shape = self.shape_interval(index).astuple()[1:]
+
+        if (
+            not index_0.subscriptable()
+            and pad_tensor
+            and index_0.value >= self.tensor_length  # type: ignore
+        ):
+            shape = self.get_empty_sample().shape
+
+        if num_samples > 0 and None in shape or self.tensor_meta.is_link:
+            sample_shapes = self._populate_sample_shapes(
+                sample_shapes,
+                index,
+                sample_shape_provider,
+                flatten=True if self.is_sequence else False,
+            )
+            if self.is_sequence:
+                num_samples = len(sample_indices)
+                try:
+                    sample_shapes = sample_shapes[np.newaxis, :].reshape(
+                        num_samples, -1, sample_ndim
+                    )
+                except ValueError:
+                    sample_shapes_list = []
+                    for i in sample_indices:
+                        start, end = self.sequence_encoder[i]
+                        sample_shapes_list.append(sample_shapes[start:end])
+                    return sample_shapes_list
+        else:
+            sample_shapes[:] = shape
+
```

Comment:
Same here, methods are too long. If you divide long function into several smaller functions, it will be much easier to test them and code will become much more readable.